flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -231,6 +234,8 @@ public void onStart() throws Exception {
         }
 
         startRecoveredJobs();
+        cleanupDirtyJobs();
```

Comment:
I'm not sure about having a separate code path for already terminated jobs as the dispatcher is already fairly complex.

What you think about unifying these and having a custom JobManagerRunner implementation that only performs checkpoint cleanup?

Reply:
I guess, that's a good point. It should be doable - it just didn't cross my mind. The `JobManagerRunner` could be in charge of handling the job's lifecycle entirely, i.e. also including the cleanup. I will work on it

Reply:
I went over the JobManagerRunner approach with Chesnay in a discussion: We addressed the issue from a user's point of view: We somehow want to visualize the running jobs and the jobs that are completed but still in cleanup phase. Having everything in `Dispatcher.runningJobs` makes sense if we want to consider jobs in cleanup mode still being part of the general job lifecycle. The `JobManagerRunner` could be responsible for maintaining this lifecycle (even after the `JobMaster` finished). That would mean that we would move the cleanup logic into the `JobManagerRunner` and have it been used by both, the `JobMasterServiceLeadershipRunner` and the new `JobManagerRunner` implementation that is in charge of cleaning up the checkpoint-related artifacts.

The flaw of this approach is, that running jobs are meant to have a `ExecutionGraphInfo` which is provided by the `JobMaster`. For the cleanup, we don't have this information. It is accessible in the `ExecutionGraphInfoStore` in the dispatcher, though. We could use that store in the new implementation. But the `ExecutionGraphInfoStore` is only persisted locally. Hence, the data would be gone in case of failover.

One workaround for that would be moving the `ExecutionGraphInfo` into the `JobResultStore` as additional metadata. Essentially, we would merge `ExecutionGraphInfoStore` and the `JobResultStore` into a single store. The issue with that approach is, that the `ExecutionGraphInfoStore` has no requirements to be backwards-compatible right now. Persisting the `ExecutionGraph` in the `JobResultStore` would change that which is not what we want, I guess. Having a `JobResultStore` that stores only a limited amount of metadata per job makes it easier to maintain backwards-compatibility.

Another approach is treating the two phases separately: `Dispatcher.runningJobs` is actually only listing jobs with a `JobMaster` providing the `ExecutionGraph`. The dirty jobs should be handled separately through a separate member (and accessed through a dedicated REST API endpoint as well).

We can move the cleanup logic into its own class to remove complexity from the `Dispatcher`.

Reply:
How is the user then informed about problems? There's going to be a dedicated REST endpoint (and new section in the Flink UI) listing the jobs from the `JobResultStore`. Any dirty jobs can be labeled as "cleanup pending". The `Dispatcher` tries to clean things up. The user is informed about issues through the logs. We could also think about adding some kind of retry counter that is persisted in the `JobResultStore`. The `Dispatcher` tries to clean things up infinitely.

Reply:
How does the user handle failures? The cluster shutdown can be either blocking until all jobs are cleaned up? Or fail with a non-0 exit code in case of jobs not being able to be cleaned up. That would trigger a restart in case of the Application Mode, which would pick up the dirty jobs again for cleanup. Here, it might make sense to add set some limit on the retries after a shutdown is triggered.

Reply:
Thinking about it once more, we could actually follow your approach by extracting a new interface out of the `JobManagerRunner` interface which does only contain `getResultFuture()`, `start()` and `getJobID()`. The `JobManagerRunner` could extend this interface. A new implementation of the extracted interface could be used to implement the checkpoint-related cleanup. After that is done, the common cleanup other components can be triggered. We could add a method `cancelCleanup()` to the new interface to enable explicit cancelling of the cleanup phase. Chesnay voted against reusing `JobManagerRunner.cancel()` because of different semantics (cancelling the job resulting in a cancelled job vs. cancelling the job cleanup resulting in a still globally-terminated job).

That would enable us to cancel the cleanup of a single job without shutting down the cluster.